tests/lib/rules/no-array-prototype-extensions.js

```diff
@@ -0,0 +1,276 @@
+'use strict';
+
+const rule = require('../../../lib/rules/no-array-prototype-extensions');
+const RuleTester = require('eslint').RuleTester;
+
+const { ERROR_MESSAGE } = rule;
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({
+  parser: require.resolve('@babel/eslint-parser'),
+  parserOptions: { ecmaVersion: 2020, sourceType: 'module' },
+});
+
+ruleTester.run('no-array-prototype-extensions', rule, {
+  valid: [
+    '[1, 2, 4].filter(el => { /* ... */ })',
+    'something.filter(el => { /* ... */ })',
+    "filterBy(something, 'foo')",
+    'this.filterBy()',
+    'something[0]',
+    'something[0].foo',
+    "get(something, 'foo.0.bar')",
+    "get(something, 'foo.0')",
+    "get(something, 'foo')[0]",
+    "get(something, 'firstObjects')",
+    "get(something, 'lastObjects')",
+    'something.length',
+    'something[something.length - 1]',
+    /** Replace is part of string native prototypes */
+    "'something'.replace()",
+    'something.replace()',
+    'something.isAny',
+    "something['compact']",
+    /** Allow directly using Ember A() */
+    `
+     import { A } from '@ember/array';
+     something = A();
+     something.filterBy('foo');
```

Comment:
It looks like this rule is currently bailing out / ignoring the entire file if `import { A } from '@ember/array';` is present. I'm not sure I agree with that.

I think what you're trying to do is allow arrays that are specifically defined as Ember arrays like this to use Ember array functions:

```js
import { A } from '@ember/array';
const something = A();
const foo = something.filterBy('foo');
```



But the problem is, this causes a lot of false negatives with any normal arrays that are using Ember array functions in the same file:

```js
import { A } from '@ember/array';
const something = A();
const foo = [];
foo.filterBy('foo');
```

For this to work, we would actually need to track the `something` variable name and only ignore Ember array functions calls on `something`. That could be done but it's more complicated as it requires tracking variables and scope.

And the current hueristic doesn't completely solve the problem anyway as there could be code that specifically expects an Ember array but without `import { A } from '@ember/array';` in the same file.

```js
function foo (anEmberArray) {
  anEmberArray.filterBy('foo'); // should be allowed since array was initialized as A() in other file
}
```

So I actually think it would be simpler and best to just avoid checking for `import { A } from '@ember/array';`. To me, that's an edge case since people rarely create arrays like that.

Reply:
Agreed. Reason I am a bit anxious about not checking for `import { A } from '@ember/array'` is that currently, it's still a recommendation in the official doc: https://guides.emberjs.com/release/configuring-ember/disabling-prototype-extensions/#toc_arrays

Reply:
Removed the checking for `@ember/array`.
Reasons:
1. As you mentioned, there are a lot of false negatives and uncovered cases.
2. Since we are moving in the direction of native functions, ember.A() would also be discouraged. Eventually, the [guide session](https://guides.emberjs.com/release/configuring-ember/disabling-prototype-extensions/#toc_arrays) can be removed. 
3. If users indeed need to use ember array, they will just have to disable the rule.